apollo-router/src/plugins/cache/entity.rs

```diff
@@ -536,45 +544,96 @@ fn filter_representations(
 // fill in the entities for the response
 async fn insert_entities_in_result(
     entities: &mut Vec<Value>,
+    errors: &[Error],
     cache: &RedisCacheStorage,
     cache_control: CacheControl,
     result: &mut Vec<IntermediateResult>,
-) -> Result<Vec<Value>, BoxError> {
+) -> Result<(Vec<Value>, Vec<Error>), BoxError> {
     let ttl: Option<Duration> = cache_control
         .ttl()
         .map(|secs| Duration::from_secs(secs as u64));
 
     let mut new_entities = Vec::new();
+    let mut new_errors = Vec::new();
 
     let mut inserted_types: HashMap<String, usize> = HashMap::new();
     let mut to_insert: Vec<_> = Vec::new();
-    let mut entities_it = entities.drain(..);
+    let mut entities_it = entities.drain(..).enumerate();
 
     // insert requested entities and cached entities in the same order as
     // they were requested
-    for IntermediateResult {
-        key,
-        typename,
-        cache_entry,
-    } in result.drain(..)
+    for (
+        new_entity_idx,
+        IntermediateResult {
+            key,
+            typename,
+            cache_entry,
+        },
+    ) in result.drain(..).enumerate()
     {
         match cache_entry {
-            Some(v) => new_entities.push(v.data),
+            Some(v) => {
+                new_entities.push(v.data);
+                for mut error in v.errors {
+                    if let Some(path) = error.path.take() {
+                        error.path = Some(
+                            Path(vec![
+                                PathElement::Key("_entities".to_string()),
+                                PathElement::Index(new_entity_idx),
+                            ])
+                            .join(path),
+                        );
+                    }
+                    new_errors.push(error);
+                }
```

Comment:
I'm wondering if it would not be more efficient to filter_map on v.errors and then extend new_errors directly to avoid re allocation of the new_errors vec. I might be wrong